plugins/heroku/src/tasks/production.ts

```diff
@@ -22,23 +22,6 @@ export default class HerokuProduction extends Task<typeof HerokuSchema> {
 
       const { scaling } = this.options
 
-      if (!scaling) {
-        const error = new ToolKitError('your heroku pipeline must have its scaling configured')
-        error.details = `configure it in your ${styles.filepath(
-          '.toolkitrc.yml'
-        )}, with a quantity and size for each production app, e.g.:
-
-options:
-  '@dotcom-tool-kit/heroku':
-    scaling:
-      ft-next-static-eu:
-        web:
-          size: standard-1x
-          quantity: 2`
-
-        throw error
-      }
```

Comment:
**question:** Do we now lose this level of plugin specific detail in the errors that have been abstracted away from the individual plugin tasks? Why did you make that decision?

Reply:
These errors will now never be hit because we do option validation when loading the plugins rather than having to check the options in the `run()` method. Previously you had to remember that even though the fields in the schema weren't optional – and therefore the given option must be specified for your Tool Kit configuration to be valid – we didn't actually do any checks when loading your configuration. So even though the options type could say an option field like `importantId` was of type `string`, as opposed to `string | undefined`, it could actually still be undefined if a user passed invalid config. So it was prone to semantic errors if you forgot to write specific checks.

As to lacking specific detail, using zod we can produce an error like
```
Options are defined in your Tool Kit configuration that are the wrong types:

- @dotcom-tool-kit/next-router has the issue(s): Expected string, received number at "appName"
- @dotcom-tool-kit/example has the issue(s): Required at "importantId"

Please update the options so that they are the expected types.
```
Do you think this is sufficient for telling users what an issue is and how they can resolve it?

Reply:
Ok I'm sold! Out of interest... how would the Heroku scaling error look as that's a funky nested object? the previous error was quite descriptive about what was required and I'm interested how Zod handles that

Reply:
Some example Heroku errors with zod:
* if you pass an empty object:
```
@dotcom-tool-kit/heroku has the issue(s): Required at "pipeline"; Required at "systemCode"; Required at "scaling"
```
* when you forget to include the process type:
```
@dotcom-tool-kit/heroku has the issue(s): Expected object, received number at "scaling.sandbox-eu.size"; Expected object, received number at "scaling.sandbox-eu.quantity"
```
* when you mix up the size and quantity fields:
```
@dotcom-tool-kit/heroku has the issue(s): Expected string, received number at "scaling.sandbox-eu.big.size"
```

And whilst we checked for the `pipeline` and `systemCode` fields before and gave nice errors before, we [didn't actually validate](https://github.com/Financial-Times/dotcom-tool-kit/blob/18e4a271c43711a1160faed4662f41b4c5329880/plugins/heroku/src/tasks/production.ts#L42-L50) the `scaling` object at all and assumed users followed the documentation exactly, which would result in confusing JavaScript's `TypeError`s!

Reply:
Thanks for getting the additional error info. These errors are not as readable as the previous error - if you were new to Tool Kit I think you would immediately know what to change if you saw the old error, but you wouldn't in some cases with the new error (I'm particularly thinking about the nested object nature of the scaling option).

But I also accept that the old error didn't cover every case and this Heroku option is an outlier in terms of its complexity so the new error is more useful in most cases. My suggestion is a litttle bit more direction for the user in the error with an additional line something like "please refer to the readme for this plugin for the full config required for this plugin". (Again I see pros and cons to that - the most definitive place to refer a user to is the schema file itself but I think the readme is better in terms of readability, even if we are having to assume that it will be kept up to date with the schema). What do you think? 😬 😁